keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1538 @@
+
+### Changes from the iptables kube-proxy backend
+
+Switching to a new backend which people will have to opt into gives us
+the chance to break backward-compatibility in various places where we
+don't like the current iptables kube-proxy behavior.
+
+#### Localhost NodePorts
+
+Kube-proxy in `iptables` mode supports NodePorts on `127.0.0.1` (but
+not `::1`) by default, which requires enabling the `route_localnet`
+sysctl, which potentially introduces security holes. (Localhost
+nodeports can be disabled via `--iptables-localhost-nodeports=false`,
+but that can't be made the default for backward-compatibility
+reasons.)
+
+(NodePorts on `::1` cannot be made to work via the normal NAT flow.
+Also, kube-proxy in `ipvs` mode does not, and cannot, implement
+localhost NodePorts even for IPv4.)
+
+We would like to avoid setting `route_localnet`. One way to do that
+would be to not support localhost NodePorts, but another way would be
+to find a way to make localhost NodePorts work without
+`route_localnet`:
+
+```
+<<[UNRESOLVED necessity of route_localnet ]>>
+
+  - In https://github.com/kubernetes/kubernetes/issues/90236#issuecomment-622355484
+    @aojea pointed out that the NAT-based approach would work for
+    IPv6 if the client explicitly set a non-localhost source IP.
+    (Presumably it would also work in the IPv4 case without needing
+    `route_localnet` in that case.) By the time the packet has reached
+    the network stack though, it is probably too late to change the
+    source IP? (OTOH, using eBPF to intercept the `connect()` call
+    would probably work.)
+
+  - It is possible that using the nftables `fwd` command from a rule
+    on a table in the `netdev` family would let us move the packet
+    around in a way that bypassed the kernel checks that otherwise
+    make this not work?
+
+  - We could resurrect enough of the userspace proxy to use it just
+    for localhost NodePort connections... :vomiting_face:
+
+      - (Alternatively: having a listening port on localhost, but then
+        after accepting a connection, using eBPF to splice it directly
+        to the destination server socket.)
+
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED localhost nodeport use cases ]>>
+
+It's not clear what the use cases for localhost NodePorts are. The
+initial implementation of NodePorts in the iptables proxy
+(https://github.com/kubernetes/kubernetes/pull/12764) set
+`route_localnet`, so localhost nodeports seem to have been considered
+an expected feature already at that point.
+
+  - In https://github.com/kubernetes/kubernetes/issues/90236#issuecomment-622142806
+    @thockin commented that one use case is connecting to a docker
+    registry without needing to set up certificates; docker
+    special-cases `localhost` specifically as not needing TLS. (@uablrek
+    also mentions in https://github.com/kubernetes/kubernetes/pull/92938#issuecomment-742985382
+    that this is "the only use-case I have heard".)
```

Comment:
Not just for certificates: it's useful because we want the container runtime on the nodes to pull from it.

So I run a nodePort service pointed at the registry running inside the cluster, and then I pull the image on any node with localhost:port

The container runtime itself can't resolve in-cluster DNS so using localhost nodeports solves this pretty elegantly at the moment.

We use this in SIG Testing for development type purposes, where the image never needs to be pushed to a remote host.

I don't know how widespread this is elsewhere, and I'm not saying this should be a prioritized use-case, but this seems difficult to reasonably implement without localhost nodeport. I suppose we could run a hostPort proxy daemonset ... 

EX: https://kind.sigs.k8s.io/docs/user/local-registry/ (where certificates aren't really the concern, though that is also handy)

Reply:
If the only thing you need is "has a statically-known IP" then there are other options. By default you can reach a NodePort service on _any_ local IP, so if you know of another IP that is present (eg, `172.17.0.1` for `docker0`), you can access the service on that IP instead. (It looks like `10.244.0.1` should work in the default kind configuration?)

Alternatively, you could configure the service with an `externalIPs` and use that IP to access the service. (Ideally you'd pick an IP from a private range that wasn't already in use, but failing that you could just do `externalIPs: ["1.2.3.4"]` or whatever...)

Reply:
VIPs implemented by kube-proxy won't work for "in cluster registrY" use-case AFAIK because the container runtime on the node has to access it (containerd/cri-o/...) as opposed to a non-hostNetwork pod (?)

NodePort being bound to loopback however is reachable from the container runtime on each node in the host network.

Reply:
(as discussed in the linked issue, the container runtime _does_ have access to the service CIDR)

anyway, this section has been rewritten since these comments